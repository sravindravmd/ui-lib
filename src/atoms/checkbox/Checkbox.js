import React, { useState, useRef } from 'react'
import { v4 as uuidv4 } from 'uuid'
import PropTypes from 'prop-types'

import { StyledCheckboxWrapper, StyledText, StyledBox } from './Checkbox.styles'
import { Checkmark } from '../../assets/images'

export const Checkbox = ({
  initialState,
  label,
  name,
  value,
  onChange,
  tabIndex,
  disabled,
  className,
  ...props
}) => {
  const [selected, setSelected] = useState(initialState)
  const checkboxRef = useRef(uuidv4())

  const handleCheck = e => {
    setSelected(e.target.checked)
    onChange(e)
  }

  return (
    <StyledCheckboxWrapper
      tabIndex={tabIndex}
      disabled={disabled}
      className={className}
    >
      <StyledBox selected={selected}>
        <Checkmark opacity='0' />
      </StyledBox>
      <input
        {...props}
        type='checkbox'
        name={name || checkboxRef.current}
        onChange={handleCheck}
        value={value}
        checked={selected}
        data-testid='checkbox'
      />
      <StyledText>{label}</StyledText>
    </StyledCheckboxWrapper>
  )
}

Checkbox.defaultProps = {
  initialState: false,
  tabIndex: 0,
  disabled: false,
  className: '',
  onChange: () => {},
}

Checkbox.propTypes = {
  label: PropTypes.string.isRequired,
  value: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  name: PropTypes.string,
  initialState: PropTypes.bool,
  disabled: PropTypes.bool,
  tabIndex: PropTypes.number,
  className: PropTypes.string,
}
