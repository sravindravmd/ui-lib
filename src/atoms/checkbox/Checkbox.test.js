import React from 'react'
import { render, fireEvent, cleanup } from '@testing-library/react'
import { Checkbox } from './Checkbox'

afterEach(cleanup)

jest.mock('uuid', () => {
  return {
    v4: jest.fn(() => 1),
  }
})

describe('<Checkbox/>', () => {
  it('should render checkbox with default props', () => {
    const { container } = render(<Checkbox label='Name' value='John Doe' />)
    expect(container).toMatchSnapshot()
  })

  it('should render checkbox with external prop values', () => {
    const { container, getByTestId } = render(
      <Checkbox
        label='Name'
        name='name'
        value='John Doe'
        onChange={() => {}}
        tabIndex={1}
      />
    )
    const checkbox = getByTestId('checkbox')
    fireEvent.click(checkbox)
    expect(checkbox).toBeChecked()
    expect(container).toMatchSnapshot()

    fireEvent.click(checkbox)
    expect(checkbox).not.toBeChecked()
    expect(container).toMatchSnapshot()
  })

  it('should render disabled checkbox', () => {
    const { container } = render(
      <Checkbox
        label='Name'
        name='name'
        value='John Doe'
        initialState={true}
        onChange={() => {}}
        disabled={true}
      />
    )
    expect(container).toMatchSnapshot()
  })
})
