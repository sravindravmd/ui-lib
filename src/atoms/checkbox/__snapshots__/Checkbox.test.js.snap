// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox/> should render checkbox with default props 1`] = `
<div>
  <label
    class="sc-bdfBwQ iTTCiC"
    tabindex="0"
  >
    <span
      class="sc-dlfnbm gspxga"
    >
      <svg
        fill="none"
        height="8"
        opacity="0"
        viewBox="0 0 11 8"
        width="11"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M1.28033 3.94163C0.987437 3.64874 0.512563 3.64874 0.21967 3.94163C-0.0732233 4.23452 -0.0732233 4.7094 0.21967 5.00229L2.94202 7.72464C3.23491 8.01754 3.70979 8.01754 4.00268 7.72464L10.447 1.28033C10.7399 0.987437 10.7399 0.512563 10.447 0.21967C10.1541 -0.0732233 9.67923 -0.0732233 9.38634 0.21967L3.47235 6.13365L1.28033 3.94163Z"
          fill="white"
        />
      </svg>
    </span>
    <input
      data-testid="checkbox"
      name="1"
      type="checkbox"
      value="John Doe"
    />
    <span
      class="sc-gsTCUz hTKKtX"
    >
      Name
    </span>
  </label>
</div>
`;

exports[`<Checkbox/> should render checkbox with external prop values 1`] = `
<div>
  <label
    class="sc-bdfBwQ iTTCiC"
    tabindex="1"
  >
    <span
      class="sc-dlfnbm cyOWeF"
    >
      <svg
        fill="none"
        height="8"
        opacity="0"
        viewBox="0 0 11 8"
        width="11"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M1.28033 3.94163C0.987437 3.64874 0.512563 3.64874 0.21967 3.94163C-0.0732233 4.23452 -0.0732233 4.7094 0.21967 5.00229L2.94202 7.72464C3.23491 8.01754 3.70979 8.01754 4.00268 7.72464L10.447 1.28033C10.7399 0.987437 10.7399 0.512563 10.447 0.21967C10.1541 -0.0732233 9.67923 -0.0732233 9.38634 0.21967L3.47235 6.13365L1.28033 3.94163Z"
          fill="white"
        />
      </svg>
    </span>
    <input
      data-testid="checkbox"
      name="name"
      type="checkbox"
      value="John Doe"
    />
    <span
      class="sc-gsTCUz hTKKtX"
    >
      Name
    </span>
  </label>
</div>
`;

exports[`<Checkbox/> should render checkbox with external prop values 2`] = `
<div>
  <label
    class="sc-bdfBwQ iTTCiC"
    tabindex="1"
  >
    <span
      class="sc-dlfnbm gspxga"
    >
      <svg
        fill="none"
        height="8"
        opacity="0"
        viewBox="0 0 11 8"
        width="11"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M1.28033 3.94163C0.987437 3.64874 0.512563 3.64874 0.21967 3.94163C-0.0732233 4.23452 -0.0732233 4.7094 0.21967 5.00229L2.94202 7.72464C3.23491 8.01754 3.70979 8.01754 4.00268 7.72464L10.447 1.28033C10.7399 0.987437 10.7399 0.512563 10.447 0.21967C10.1541 -0.0732233 9.67923 -0.0732233 9.38634 0.21967L3.47235 6.13365L1.28033 3.94163Z"
          fill="white"
        />
      </svg>
    </span>
    <input
      data-testid="checkbox"
      name="name"
      type="checkbox"
      value="John Doe"
    />
    <span
      class="sc-gsTCUz hTKKtX"
    >
      Name
    </span>
  </label>
</div>
`;

exports[`<Checkbox/> should render disabled checkbox 1`] = `
<div>
  <label
    class="sc-bdfBwQ dHzmZK"
    disabled=""
    tabindex="0"
  >
    <span
      class="sc-dlfnbm cyOWeF"
    >
      <svg
        fill="none"
        height="8"
        opacity="0"
        viewBox="0 0 11 8"
        width="11"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M1.28033 3.94163C0.987437 3.64874 0.512563 3.64874 0.21967 3.94163C-0.0732233 4.23452 -0.0732233 4.7094 0.21967 5.00229L2.94202 7.72464C3.23491 8.01754 3.70979 8.01754 4.00268 7.72464L10.447 1.28033C10.7399 0.987437 10.7399 0.512563 10.447 0.21967C10.1541 -0.0732233 9.67923 -0.0732233 9.38634 0.21967L3.47235 6.13365L1.28033 3.94163Z"
          fill="white"
        />
      </svg>
    </span>
    <input
      checked=""
      data-testid="checkbox"
      name="name"
      type="checkbox"
      value="John Doe"
    />
    <span
      class="sc-gsTCUz hTKKtX"
    >
      Name
    </span>
  </label>
</div>
`;
